src/jwt/mod.rs

```diff
@@ -20,11 +20,98 @@ new_type![
+        ///
+        /// See [RFC 2045 section 5.1](https://tools.ietf.org/html/rfc2045#section-5.1) for the full Content-Type Header Field spec.
+        /// See [RFC 7515 section 4.19](https://tools.ietf.org/html/rfc7515#section-4.1.9) for specific requirements of JSON Web Token Types.
+        pub fn normalize(&self) -> Result<NormalizedJsonWebTokenType, JsonWebTokenTypeNormalizeError> {
+            self.try_into()
+        }
+    }
+];
+
+new_type![
+    /// JSON web token type field (typ)
+    ///
+    /// This types stores the raw (deserialized) value.
+    /// To retrieve a normalized value according to RFC2045 and RFC7515 see [`JsonWebTokenType::normalize`]
+    ///
+    /// See [RFC 2045 section 5.1](https://tools.ietf.org/html/rfc2045#section-5.1) for the full Content-Type Header Field spec.
+    /// See [RFC 7515 section 4.1.9](https://tools.ietf.org/html/rfc7515#section-4.1.9) for specific requirements of JSON Web Token Types.
+    ///
+    /// It is recommended to only instantiate `NormalizedJsonWebTokenType` objects via [`JsonWebTokenType`] and then call [`JsonWebTokenType::normalize`].
+    ///
+    /// ```rust
+    /// # use openidconnect::{NormalizedJsonWebTokenType, JsonWebTokenType};
+    /// let token_type: NormalizedJsonWebTokenType = JsonWebTokenType::new("jwt+at".to_string())
+    ///     .normalize()
+    ///     .unwrap();
+    ///
+    /// // normalized value looks like this
+    /// assert_eq!(*token_type, "application/jwt+at")
+    /// ```
+    ///
+    /// When you really want to create this type directly, please check the following items:
+    ///
+    /// * Pass allowed types as an Iterator while always using the content type spec of https://tools.ietf.org/html/rfc2045#section-5.1
+    ///   (always prepend `application/`, e.g. you have to pass `application/jwt` instead of just `jwt`.)
+    /// * Make sure to only use lowercase characters
+    #[derive(Deserialize, Hash, Ord, PartialOrd, Serialize)]
+    NormalizedJsonWebTokenType(String)
 ];
 
+impl From<NormalizedJsonWebTokenType> for JsonWebTokenType {
+    fn from(t: NormalizedJsonWebTokenType) -> JsonWebTokenType {
+        JsonWebTokenType::new(t.0)
+    }
+}
+
+impl TryFrom<&JsonWebTokenType> for NormalizedJsonWebTokenType {
+    type Error = JsonWebTokenTypeNormalizeError;
+
+    /// Normalizes a [`JsonWebTokenType`] and produces a [`NormalizedJsonWebTokenType`] according to RFC2045.
+    ///
+    /// See [RFC 2045 section 5.1](https://tools.ietf.org/html/rfc2045#section-5.1) for the full Content-Type Header Field spec.
+    /// See [RFC 7515 section 4.19](https://tools.ietf.org/html/rfc7515#section-4.1.9) for specific requirements of JSON Web Token Types.
+    fn try_from(t: &JsonWebTokenType) -> Result<NormalizedJsonWebTokenType, Self::Error> {
+        let lowercase_jwt_type = t.0.to_lowercase();
+        if lowercase_jwt_type.contains('/') {
+            if lowercase_jwt_type.starts_with("application/") {
```

Comment:
where does the spec require MIME types to be `application/*`? I see where that's implied when omitted, but I don't see why some custom app couldn't have its own complex MIME type with a different top-level type

Reply:
good catch!

I changed the logic to match the standard now (hopefully ^^).

Also added some tests.